### ISSUE TITLE
`<shared_mutex>`: `SRWLOCK` can grant exclusive ownership when we requested shared ownership

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#4448]
Original bug report: https://www.reddit.com/r/cpp/comments/1b55686/maybe_possible_bug_in_stdshared_mutex_on_windows/

Reported to Windows as internal OS-49268777 "`SRWLOCK` can deadlock after an exclusive owner has released ownership and several reader threads are attempting to acquire shared ownership together".

<details><summary>Click to expand reduced STL repro:</summary>

```
C:\Temp>type meow.cpp
```
```cpp
#include <atomic>
#include <cstdio>
#include <shared_mutex>
#include <thread>
#include <vector>

constexpr int numThreads = 5;

std::shared_mutex g_shared_mutex;

std::atomic<int> g_atomicReadCounter{0};

void DoStuff() {
    g_shared_mutex.lock_shared();

    // Spin and yield until all reader threads have acquired their shared lock
    g_atomicReadCounter.fetch_add(1);
    while (g_atomicReadCounter.load() != numThreads) {
        std::this_thread::yield();
    }

    g_shared_mutex.unlock_shared();
}

int main() {
    std::vector<std::jthread> readerThreads;

    for (int loop = 0; true; ++loop) {
        g_atomicReadCounter.store(0);

        g_shared_mutex.lock();

        for (int i = 0; i < numThreads; ++i) {
            readerThreads.emplace_back(DoStuff);
        }

        g_shared_mutex.unlock();

        readerThreads.clear(); // Join all threads

        std::printf("If this stops increasing, we've deadlocked: %d\n", loop);
    }
}
```
```
C:\Temp>cl /EHsc /nologo /W4 /std:c++20 /MTd /Od meow.cpp
meow.cpp

C:\Temp>meow
If this stops increasing, we've deadlocked: 0
If this stops increasing, we've deadlocked: 1
[...]
If this stops increasing, we've deadlocked: 1814
If this stops increasing, we've deadlocked: 1815
```
</details>

This code is squirrelly :chipmunk: in that it's trying to get all of the worker threads to share ownership simultaneously (instead of using `shared_mutex` for its usual purpose, where exclusive owners want to write, shared owners want to read, and everyone wants to relinquish ownership as fast as possible). However, my reading of the Standard indicates that this is conforming code, because (1) nothing says that requesting shared ownership can actually grant exclusive ownership, and (2) WG21-N4971 \[thread.sharedmutex.requirements.general\]/2 says:

> The maximum number of execution agents which can share a shared lock on a single shared mutex type is unspecified, but is at least 10000. If more than the maximum number of execution agents attempt to obtain a shared lock, the excess execution agents block until the number of shared locks are reduced below the maximum amount by other execution agents releasing their shared lock.

I read this as saying that a conforming implementation must allow at least 10,000 clowns to pile into the clown car here :clown_face: :car:. Instead, when this repro happens, we effectively have a limit of 1.

@adr26 reached out to me, as he worked on the `SRWLOCK` implementation before leaving Microsoft, and confirmed that this behavior, while not documented on Microsoft Learn, is a consequence of a performance optimization. @cpkleynhans also confirmed this in OS-49268777.

Because of this mismatch between the C++ Standard's requirements and `SRWLOCK`'s behavior, Windows will need to decide what to do before the STL potentially takes action:

* If Windows changes `SRWLOCK`'s behavior, then no action is required from the STL.
  + Affected users would need to upgrade their OS.
* If Windows provides some way for the STL to request actually-shared ownership, then we'd need to write code to use that mechanism.
  + It would need to be conditionally called (because the STL cannot assume that it's running on the latest OS), which we have the machinery for. This would impose relatively minor performance costs.
* If there's a way for us to detect that we got exclusive ownership when we requested shared, we could unlock and try again.
  + That would presumably involve inspecting `SRWLOCK`'s representation (a major maintenance/future-compatibility burden), and would have major performance impacts.
  + @adr26 says that it's almost certainly physically impossible to distinguish this scenario from other scenarios.
* If Windows hypothetically allowed it, and the code doesn't involve any Windows-internal API calls, we could potentially replicate the shared lock logic into the STL itself, but without the performance optimization that causes these shared => exclusive upgrades.
  + This would also impose maintenance and future-compatibility burdens, but wouldn't impact performance beyond giving up the optimization that we don't want.
  + Long ago, we intended to do something very similar for simulating `TryAcquireSRWLockExclusive` (a Win7 API) when running on Windows Vista, in our stranded "WCFB02" branch.
* In vNext, we could entirely replace `shared_mutex`'s implementation.
  + This would significantly increase our maintenance burden (implementing a thin wrapper is easy, implementing actual logic is hard, as with `shared_timed_mutex`), and would presumably significantly impact performance. The whole purpose of `shared_mutex` is to be faster than `mutex` (allowing sharing) and `shared_timed_mutex` (giving up the timing functionality).
* We could push for a change in the Standard itself to reflect our now-longstanding implementation behavior.
  + This would presumably need a small paper, not an LWG issue. I think I'd want to change "but is at least 10000" to "but is at least 1".
  + @adr26 implemented this `SRWLOCK` optimization circa 2011, we shipped `shared_mutex` in VS 2015 Update 2 released on 2016-03-30, and the world hasn't melted yet.
  + The rationale would be that this shared => exclusive upgrade behavior is actually a consequence of a performance optimization, and does not affect *intended* use of `shared_mutex`. Users who want a bunch of threads to all gather before doing other work should be using other Standard mechanisms like `barrier` or `latch`.
* We could do nothing.
  + We *really* don't like to wontfix STL bugs, though. Conformance issues are eternal.

My preferred options (unsurprisingly) are for Windows to change `SRWLOCK`'s behavior, or for the Standard to change its requirements, so that no STL changes will be needed.

**Comment by adr26** [comment]
Two quick comments on the above:

* NeillC originally implemented the original 2-phase unlock algorithm which can allow the shared->exclusive upgrade path some time after pushlocks were first created, but before Win7 (I can't remember how far back it was). I then expanded the use of 2-phase unlock to more cases in 2011. In 2011, I made hitting this more likely, but the risk had been there previously as well.
   
* I think that for internal reasons to do with the SRW implementation, it wouldn't be possible to discriminate between a lock which was acquired using a shared acquire API having gone through the shared->exclusive upgrade path, and when it was acquired without contention but then waiters arrived on the lock subsequently. Assuming that's correct (as I don't have sight of the history since I left), it would preclude STL being able to inspect the SRW state as a means of detecting this.

**Comment by AlexGuteniev** [comment]
I would not want change to the standard.

I don't beleive that not alowing multiple shared locks can be a real optimiation. This might optimize some cases, but is a disastrous pessimization for those where shared mutex is needed most.

**Comment by AlexGuteniev** [comment]
> * there's a way for us to detect that we got exclusive ownership when we requested shared, we could unlock and try again.
>   
>   * That would presumably involve inspecting `SRWLOCK`'s representation (a major maintenance/future-compatibility burden), and would have major performance impacts.
>   * @adr26 says that it's almost certainly physically impossible to distinguish this scenario from other scenarios.

I think there's a way: try to acquire the same `SRWLOCK` again using `TryAcquireSRWLockShared`, if it fails, we're in this state. Sure this will alert verification tools that makes sure `SRWLOCK` is not used recursively. And will need to release double. But otherwise should work, I think

**Comment by StephanTLavavej** [comment]
Trying to recursively acquire shared ownership seems extremely squirrely and is not something I'd want to do in production code.

**Comment by NeillClift** [comment]
The reason this path exists is because we want to avoid convoys by unconditionally releasing the lock immediately and then having one thread find threads to wake. Any convoy would be broken by the lock getting stolen by other threads trying to acquire. For the exclusive case this is just a bit test and set. When I implemented the shared acquire, there isn't enough state to record this. Since shared acquires aren't generally compatible with other shared acquires and the emphasis is on getting threads through fast (that whole convoy avoidance idea) I thought this was OK and important.
Now that I think on this again I realize that we have the waker active and we can just queue a wait block to the lock in this case and the active waker will sort this all out. Obviously, there is retry for this whole thing just like all the paths as stuff can change.
We the delayed acquisition lead to some performance problem? Well, the running shared acquire might not be able to get the lock if it's put behind an exclusive acquirer who needs to context swap in. Any reasonable usage though should have a high number of readers though.

**Comment by adr26** [comment]
I agree with @NeillClift: a shared acquirer that sees the lock bit free and waiters to be woken could push itself on the waitlist and be woken to make a shared acquire at some point in the future, without changing the way the SRWLOCK works otherwise. In particular, that means that exclusive acquirers could still steal the lock if the lock bit is clear and there are waiters present, to help break convoys.

If it was my call and a change to the C++ spec to allow the current behaviour were ruled out, then I'd suggest that a shared acquire API that had this behaviour was added to Win32 SRWLOCKs, with the original shared acquire API kept as-is, to give it maximum convoy-busting ability (i.e. minimise the perf hit to only users of the new API, viz. the C++ STL).

> I don't believe that not allowing multiple shared locks can be a real optimization.

Lock convoys are a real and very serious performance problem, and are well known in academic literature. The seminal paper describing them, written by a group including Jim Gray, dates from 1977 and can be read here:

https://dl.acm.org/doi/pdf/10.1145/850657.850659

> I have an example in codebase where `shared_mutex` is used to protect a resource (i.e. intended use), but still shared acquisition is made for potentially long time, and preventing it is likely to cause bad responsiveness or even deadlocks. Will have to replace that with `boost::` alternative.

It would be informative to be able to profile your code under heavy load on a large MP machine: though you are correct that promoting a shared acquirer -> exclusive must needs entails that that shared acquirer precludes concurrent access by other shared acquirers, there likely is a performance benefit of even a potentially long shared lock acquisition being able to go ahead immediately rather than maintaining a convoy (cf. the paper). Windows is a large codebase with many different users of this lock (both user and kernel mode code), and the result for it of @NeillClift making this change was a clear performance improvement. Though potentially counterintuitive, this may be the case for your code too &mdash; the best thing to do is measure and check!

**Comment by AlexGuteniev** [comment]
> It would be informative to be able to profile your code under heavy load on a large MP machine

I think that's inapplicable. It is not something computational.

The resource being managed is some external hardware connected via network or ports, the excusive lock is some sort of background action on timer, the shared locks are acquired and released by user's action to open and close some windows. Don't mind if exclusive lock isn't acquired when user is doing something and one of those windows are open. Still not showing a window and hanging when another is open is a bug.

Sure if it was documented for `std::shared_mutex` can block shared acquisitions sometimes (for performance reasons or whatever), I'd implement my own primitive for that, but the C++ documents contrary.

Maybe my requirements are not common. In this project we already do things that are considered questionable. Specifically, we have `fair_mutex` for places where fairness is more important than throughput.

**Comment by AlexGuteniev** [comment]
> maximum convoy-busting ability

Ok, I now see the rationale. Since it looks like we are in the current state for quite long (there would be a C++ paper push, which would be reviewed), would you mind documenting the current behavior? Unfairness and some other internals are already documented, maybe not next to the functions, but [here](https://learn.microsoft.com/en-us/windows/win32/Sync/slim-reader-writer--srw--locks).

**Comment by NeillClift** [comment]
If I was making the change I would just change the shared acquire to avoid this stealing path. Why? because unless you have a high number of shared acquires a share/exclusive lock probably doesn't make sense. It's also less inherently important as the exclusive path. Needs to be shaken out with perf tests in windows etc.
The convoy issue was massive in the early days. Perf team would go on and on about it for years. I didn't really understand the issue properly until I met with the research folks. Once I understood the 'lock hold time expanded by context swap time' everything made sense. I don't think it's a well-known issue. Going unfair was huge in terms of performance. That would have been all the exclusive case though.
I never anticipated this rendezvous thing though.

**Comment by adr26** [comment]
@NeillClift:

This is how I see it: the `SRWLOCK` is a general-purpose synchronisation primitive provided by Windows, so should be designed to provide optimal throughput over a wide variety of uses.

If you have a pattern where there are shared acquirers holding the lock for a long average period (P<sub>s</sub>) and there are a large average number of shared acquirers (N<sub>s</sub>), then at the point an exclusive acquirer comes along and enqueues, future shared acquirers will enqueue behind the exclusive waiter but that exclusive waiter is still going to have to wait for an average number of shared acquirers to drop the lock before the exclusive access is granted.

Assuming a normal distribution of hold times for shared acquirers, on average the shared acquirers will be half way through their hold time (½P<sub>s</sub>) at the point the exclusive waiter arrives. The exclusive waiter then has to wait for the last of the shared acquirers currently holding the lock at the point it enqueues to drop the lock before it can acquire (ignoring any anti-convoy lock stealing which might expand this wait time).

If N<sub>s</sub> is large enough, then rather than hitting the middle of the normal distribution (or near it) for a sample of N<sub>s</sub> holders, you will get a sample that's more spread over the distribution of hold times &mdash; and it's the supremum of that set of hold times which determines the latency of the exclusive acquire. I realise a normal distribution is an imperfect assumption for lock hold times, as on the lower side you can't get a zero or negative lock hold time(!), but it's probably still OK the tail to think of it as a normal distribution truncated at zero, in which case the distribution on the upper side (longer lock times) is potentially bounded only by the maximum shared lock hold time (max(P<sub>s</sub>)) as N<sub>s</sub> grows... or put another way, the latency on the exclusive acquire is going to tend toward the maximum shared lock hold time (max(P<sub>s</sub>)) as N<sub>s</sub> becomes large.

Therefore, I think you can make a strong argument that long shared lock times are a usage pattern that is going to hurt performance, whether the lock has anti-convoy features or not. As such, when optimising `SRWLOCK` as a general-purpose synchronisation primitive provided by Windows, consideration should be first be given to high-performance cases whose performance can be improved by anti-convoy features, and consideration of inherently-low-performance cases like these comes second.

Everyone who's done perf work knows that it's incredibly hard to reason about what the effect of changes are going to be, so I hold the above lightly and I 100% agree with the sentiment that it 'needs to be shaken out with perf tests in windows' &mdash; i.e. that quantified performance metrics from testing are the way to evaluate these changes.

The rendezvous problem essentially arises from the C++ STL layering a requirement on top of the `SRWLOCK` that isn't specified in the `SRWLOCK` documentation, viz. that any shared acquirer of the lock must be able to admit a certain number of other shared acquirers concurrently. For `SRWLOCK` the maximum number of other shared acquirers you can rely on acquiring the lock at the same time as you has always effectively been `0`, because of the shared->exclusive upgrade path.

I agree that whatever decisions are taken, the Win32 API docs for `SRWLOCK` should be clarified to document the maximum number of concurrent shared acquirers the caller can rely on having. If the shared->exclusive upgrade path is found to still be a perf win and is kept, then whatever mechanism is added to allow a caller (such as the C++ STL) to force it to enqueue a waitblock rather than use the shared->exclusive upgrade path (and so to ensure other shared acquirers are allowed) should also be clearly documented.

@StephanTLavavej:

I have been reading this document from @tkoeppe about how the C++ standard expresses normative requirements on an implementation:

https://github.com/cplusplus/draft/wiki/Specification-Style-Guidelines#standards-wording-idioms

It says that normative requirements on an implementation are expressed using **shall** and **is**.

However, [thread.sharedmutex.requirements.general]/2 uses the word **can**. In the C++ specification style guidelines referenced above, **can** is only used in an example of replacing 'might therefore be' with 'can result', as a means of expressing an optional positive action.

Is therefore a legitimate reading of the text from [thread.sharedmutex.requirements.general]/2:

> The maximum number of execution agents which can share a shared lock on a single shared mutex type is unspecified, but is at least 10000. If more than the maximum number of execution agents attempt to obtain a shared lock, the excess execution agents block until the number of shared locks are reduced below the maximum amount by other execution agents releasing their shared lock.

that it:

1. Allows an implementation to specify a *maximum number of execution agents* which are able to share a shared lock (though does not require it to specify such a limit, if effectively the maximum number of execution agents which are able to share a shared lock is greater than or equal to the maximum number of execution agents the implementation supports).

2. It imposes a requirement on the implementation's behaviour when more than the specified *maximum number of execution agents* attempt to obtain a shared lock (i.e. that the excess execution agents block until the number of shared locks are reduced below the maximum amount by other execution agents releasing their shared lock).

3. Does not specify a requirement for a *minimum number of execution agents* which **shall** be able to concurrently share a lock but only specifies what **shall** occur if more than the specified *maximum number of execution agents* attempt to obtain a shared lock.

If the above interpretation of [thread.sharedmutex.requirements.general]/2 is valid, then the current C++ STL implementation (built on the current STL behaviour) is conforming and no change is required, either in the `SRWLOCK` API from Windows or in the C++ STL.

FWIW, I think the original motivating code from Reddit is squirrelly 🐿️, since it requires a caller holding the (shared) lock to busy-wait on an action (the increment of the shared counter) which effectively entails the acquisition of the lock again. Explicit re-entrant lock acquisition is vetoed for `SRWLOCK`, and in this case a thread holding the lock is logically waiting for the lock to be acquired again, albeit by indirect means of another thread performing that action, so in my mind it qualifies as re-entrant behaviour, which is not allowed with the `SRWLOCK`. In any case, the motivating code would be more efficient and naturally implemented using condition variables to wait on the shared counter reaching the desired value, so that the wait for the counter increment could occur without the lock being held. As well as avoiding the deadlock, does this would also minimise the lock hold time of that code. This is precisely the motivating use case for condition variables and what they're designed to do!

**Comment by AlexGuteniev** [comment]
I created MicrosoftDocs/sdk-api#1785

**Comment by StephanTLavavej** [comment]
@adr26

> It says that normative requirements on an implementation are expressed using **shall** and **is**.
> 
> However, \[thread.sharedmutex.requirements.general\]/2 uses the word **can**.

It's talking about what user code is capable of, not what implementations need to do.

Also, Library Standardese isn't usually as carefully drafted as Core Standardese. It does get cleaned up over time, but we don't really worry intensely about the distinction between "shall" and other words. From a library implementer perspective, if the Standard says something outside of a Note or an Example (both of which are always informative, not normative), the Standard is generally saying something important that either users need to follow (some restriction on their behavior) or that implementations need to support. The Library specifically calls out when it's making suggestions, nowadays usually with a "Recommended practice:" element.

> Is therefore a legitimate reading of the text from \[thread.sharedmutex.requirements.general\]/2:

I don't think so - I think you're trying really hard to read it as saying that the current `SRWLOCK` implementation maps exactly to what C++ specifies - yet we know that nobody on the Committee knew about `SRWLOCK`'s behavior here.

The Standard could certainly be phrased to distinguish a "maximum" and a "minimum" number as you suggest, but I don't think anyone was envisioning that distinction.

My reading of the current Standardese is pretty simple - in the absence of an exclusive owner, nothing says that attempting to acquire shared ownership is allowed to block indefinitely, except for this part about reaching the >=10k maximum. The situation that `SRWLOCK` creates, of "maybe you can only get 1!" is so surprising that if the Standard wanted to allow it, it should have been explicitly called out.

I agree that the user code is squirrelly, but we have to support lots of squirrelly usage. I'm not sure I agree that it's "morally re-entrant" but it is certainly not physically re-entrant and I don't think that reasoning can be used to rule out the code as conforming.

To summarize - while I would love to find a reading that says we (either Windows or the STL) don't have to do any work, I just don't see it. I think the chances of the Library Working Group agreeing that the Standardese as written allows our behavior is extremely low (1%). I also think the chances of getting a wording change through Library Evolution and the full Committee that permits our behavior is low (10%), given how contributors have reacted to this discovery.

**Comment by adr26** [comment]
Hi @StephanTLavavej,

Thank you for your description of why the C++ standard can't be considered to match the current implementation: you certainly have more expertise in the area of the C++ libraries standardisation than I do, and I appreciate you taking the time to explain why the spec couldn't be considered to allow this.

I think that @AlexGuteniev's example is a good motivator for why having an API that guarantees a shared acquire doesn't get upgraded to an exclusive acquire could be a good idea. My main concern (and the reason I was asking) was simply a concern for overall OS perf, if there isn't an option to retain the current `SRWLOCK` behaviour for callers who don't need that (which could be the majority outside the C++ STL).

Based on what you're saying, it seems like the C++ STL almost definitely needs to be able to opt into a behaviour which guarantees that a shared acquire doesn't get upgraded to an exclusive acquire. The decision of whether to make a unilateral change (and accept any perf hit) in the `SRWLOCK` API, or whether to provide a new API/option (and in that case whether to make the new behaviour opt-in or opt-out) will have to be handled by the Windows team, and then once that decision's made the STL will have to be updated accordingly (I would even be happy to offer to update it!).

In the interim, I thought it might be helpful to illustrate how condition variables can be used to avoid this. I see that the STL provides nice wrappers for the underlying Win32 condition variable APIs, so I've built an example analogous to yours, using `std::condition_variable_any` and `std::shared_mutex`, to illustrate how using condition variables should solve the problem:

<details>
  <summary>Example implementation using <code>std::condition_variable_any</code> which does not deadlock</summary>

  ```c++
  #include <atomic>
  #include <cstdio>
  #include <shared_mutex>
  #include <thread>
  #include <vector>
  
  constexpr int numThreads = 5;
  
  std::shared_mutex g_shared_mutex;
  std::condition_variable_any g_all_started;
  
  int g_read_counter;
  
  void DoStuff() {
      // Register that this thread has started
      g_shared_mutex.lock();
      g_read_counter += 1;
      g_shared_mutex.unlock();
  
      g_all_started.notify_all();
  
      // Wait until all reader threads have started
      std::shared_lock lock(g_shared_mutex);
      g_all_started.wait(lock, []{ return g_read_counter == numThreads; });
  
      // Do shared work after all threads have started
  
      // Shared work completed, drop lock on exit
  }
  
  int main() {
      std::vector<std::jthread> readerThreads;
  
      for (int loop = 0; true; ++loop) {
          g_read_counter = 0;
  
          g_shared_mutex.lock();
  
          for (int i = 0; i < numThreads; ++i) {
              readerThreads.emplace_back(DoStuff);
          }
  
          g_shared_mutex.unlock();
  
          readerThreads.clear(); // Join all threads
  
          std::printf("If this stops increasing, we've deadlocked: %d\n", loop);
      }
  }
  ```
</details>

However, when I did this, I noticed that `Concurrency::details::stl_condition_variable_win7::wait_for` is calling `SleepConditionVariableSRW` without using `CONDITION_VARIABLE_LOCKMODE_SHARED` in `Flags` (it's just always `0`!):

```c++
            bool wait_for(_Stl_critical_section* lock, unsigned int timeout) {
                return SleepConditionVariableSRW(
                           &m_condition_variable, reinterpret_cast<PSRWLOCK>(&lock->_M_srw_lock), timeout, 0)
                    != 0;
            }
```
https://github.com/microsoft/STL/blob/d6efe9416e4ad7d6e245ae9e96023d413794d1eb/stl/src/primitives.hpp#L29

```
>	mutex.exe!Concurrency::details::stl_condition_variable_win7::wait_for(_Stl_critical_section * lock, unsigned int timeout) Line 28	C++
 	mutex.exe!Concurrency::details::stl_condition_variable_win7::wait(_Stl_critical_section * lock) Line 22	C++
 	mutex.exe!_Cnd_wait(_Cnd_internal_imp_t * cond, _Mtx_internal_imp_t * mtx) Line 65	C++
 	mutex.exe!std::condition_variable_any::wait<std::shared_lock<std::shared_mutex>>(std::shared_lock<std::shared_mutex> & _Lck) Line 80	C++
 	mutex.exe!std::condition_variable_any::wait<std::shared_lock<std::shared_mutex>,`DoStuff'::`2'::<lambda_1>>(std::shared_lock<std::shared_mutex> & _Lck, DoStuff::__l2::<lambda_1> _Pred) Line 89	C++
 	mutex.exe!DoStuff() Line 29	C++
```

(cf. description of `CONDITION_VARIABLE_LOCKMODE_SHARED` in https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleepconditionvariablesrw#parameters)

[WG21-N4971](https://wg21.link/N4971) says of `std::condition_variable_any::wait` in [thread.condvarany.wait]/1.2:

> &mdash; When unblocked, calls `lock.lock()` (possibly blocking on the lock) and returns.

and [thread.lock.shared.locking]/1 says of `std::shared_lock::lock`:

> _Effects:_ As if by `pm->lock_shared()`.

Therefore, though my sample code is calling `std::condition_variable_any::wait()` with a `std::shared_lock<std::shared_mutex>`, the `std::shared_mutex` is being locked in exclusive mode on return from `SleepConditionVariableSRW()`, instead of in shared mode. This appears to me to be a bug in the current STL implementation (`CONDITION_VARIABLE_LOCKMODE_SHARED` should be used for `std::shared_lock<std::shared_mutex>`).

In my example, after checking that all the callers have started and attempting to acquire the `std::shared_mutex` in shared mode via `std::condition_variable_any::wait()` (but actually being given the lock in exclusive mode), the callers aren't doing anything which depends on them having shared access, so forcing them to have exclusive access is just a pessimisation (and fortunately doesn't cause my code to deadlock!), but on the assumption that you agree this violates the spec, I think this should be fixed so that they can all run concurrently.

Thanks, Andrew R

Thanks, Andrew R

**Comment by adr26** [comment]
I filed #4462 to track the above!

**Comment by AlexGuteniev** [comment]
> Based on what you're saying, it seems like the C++ STL almost definitely needs to be able to opt into a behaviour which guarantees that a shared acquire doesn't get upgraded to an exclusive acquire. The decision of whether to make a unilateral change (and accept any perf hit) in the `SRWLOCK` API, or whether to provide a new API/option (and in that case whether to make the new behaviour opt-in or opt-out) will have to be handled by the Windows team, and then once that decision's made the STL will have to be updated accordingly (I would even be happy to offer to update it!).

I'm strongly in favor for making the new behavior default, and only the affected by lock convoys portions Windows code should opt out. Whereas I mostly hate `shared_mutex` being unreliable, I'd also want `SRWLOCK` to be fixed. I think the code severely impacted by lock convoys should instead fix its lock usage, like reducing parallelism, reducing locks scope, going lock-free where applicable, etc, rather that relying on that the lock itself would automatically fix their problem.

**Comment by adr26** [comment]
The `SRWLOCK` code is essentially a user-mode replica of the code for `EX_PUSH_LOCK` in the kernel (the main difference being the mechanism used to cause a thread to wait when it has to block on the lock) and the two have been kept in sync over time.

I can't remember exactly when @NeillClift first introduced push locks (they were initially internal to the kernel and not exported), but I have done a bit of internet archaeology and reverse engineering and it appears that the `EX_PUSH_LOCK` APIs were first exported from the kernel in Windows Server 2003 RTM (`5.2.3790.0 (srv03_rtm.030324-2048)`).

Examining `ExfAcquirePushLockShared` from the Windows Server 2003 RTM build, I see that it had the waiters and lock bit positions reversed, the lock bit was only used for exclusive owners, and it didn't have the anti-convoy changes (e.g. the shared -> exclusive upgrade path):
```
C:\Temp>curl -Lo ntos.ex_ https://archive.org/download/WinSrv200310in1/Windows%202003%20Server%2010%20In%201%20CD.iso/23SC/NTKRNLMP.EX_
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 1051k    0 1051k    0     0  50654      0 --:--:--  0:00:21 --:--:-- 50660

C:\Temp>expand ntos.ex_ ntos.exe
Microsoft (R) File Expansion Utility
Copyright (c) Microsoft Corporation. All rights reserved.

Adding C:\Temp\ntos.exe to Extraction Queue

Expanding Files ....

Expanding Files Complete ...

C:\Temp>cdb -z ntos.exe

Microsoft (R) Windows Debugger Version 10.0.22621.755 X86
Copyright (c) Microsoft Corporation. All rights reserved.


Loading Dump File [C:\Temp\ntos.exe]

************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
Symbol search path is: srv*
Executable search path is:
ModLoad: 00400000 00666000   C:\Temp\ntos.exe
eax=00000000 ebx=00000000 ecx=00000000 edx=00000000 esi=00000000 edi=00000000
eip=00612020 esp=00000000 ebp=00000000 iopl=0         nv up di pl nz na po nc
cs=0000  ss=0000  ds=0000  es=0000  fs=0000  gs=0000             efl=00000000
ntos!SeTokenObjectType+0x13ab4:
00612020 55              push    ebp

0:000> lm vm nt*
start    end        module name
00400000 00666000   ntos       (export symbols)       ntos.exe
    Loaded symbol image file: ntos.exe
    Mapped memory image file: C:\Temp\ntos.exe
    Image path: C:\Temp\ntos.exe
    Image name: ntos.exe
    Timestamp:        Tue Mar 25 08:39:34 2003 (3E8015C6)
    CheckSum:         00254553
    ImageSize:        00266000
    File version:     5.2.3790.0
    Product version:  5.2.3790.0
    File flags:       0 (Mask 3F)
    File OS:          40004 NT Win32
    File type:        1.0 App
    File date:        00000000.00000000
    Translations:     0409.04b0
    Information from resource tables:
        CompanyName:      Microsoft Corporation
        ProductName:      Microsoft® Windows® Operating System
        InternalName:     ntkrnlmp.exe
        OriginalFilename: ntkrnlmp.exe
        ProductVersion:   5.2.3790.0
        FileVersion:      5.2.3790.0 (srv03_rtm.030324-2048)
        FileDescription:  NT Kernel & System
        LegalCopyright:   © Microsoft Corporation. All rights reserved.

0:000> uf ntos!ExfAcquirePushLockShared
ntos!ExfAcquirePushLockShared:
004c4750 55              push    ebp
004c4751 8bec            mov     ebp,esp
004c4753 83ec24          sub     esp,24h
004c4756 53              push    ebx
004c4757 56              push    esi
004c4758 8bf1            mov     esi,ecx
004c475a 8b1e            mov     ebx,dword ptr [esi]                           ; load lock value -> ebx
004c475c 33c9            xor     ecx,ecx                                       ; ecx = 0

ntos!ExfAcquirePushLockShared+0xe:
004c475e f6c303          test    bl,3                                          ; check for lock (2) or waiters (1)
004c4761 7451            je      ntos!ExfAcquirePushLockShared+0x64 (004c47b4) ; if no lock or waiters, go to +0x64

ntos!ExfAcquirePushLockShared+0x13:
004c4763 f6c301          test    bl,1                                          ; check for waiters (1)
004c4766 8d45e4          lea     eax,[ebp-1Ch]
004c4769 c645dc01        mov     byte ptr [ebp-24h],1                          ; set up wait block @ [ebp-24h]
004c476d c645de04        mov     byte ptr [ebp-22h],4                          ; NB: KMUTEX at start of block / mutex init'd for wait
004c4771 894de0          mov     dword ptr [ebp-20h],ecx
004c4774 8945e8          mov     dword ptr [ebp-18h],eax
004c4777 8945e4          mov     dword ptr [ebp-1Ch],eax
004c477a 884dfc          mov     byte ptr [ebp-4],cl
004c477d 894df8          mov     dword ptr [ebp-8],ecx
004c4780 894df0          mov     dword ptr [ebp-10h],ecx
004c4783 894df4          mov     dword ptr [ebp-0Ch],ecx
004c4786 7427            je      ntos!ExfAcquirePushLockShared+0x5f (004c47af) ; if no waiters (1), => locked (2) set, go to +0x5f

ntos!ExfAcquirePushLockShared+0x38:
004c4788 8d43ff          lea     eax,[ebx-1]                                   ; clear waiters bit (1) to obtain waiters list ptr
004c478b 8945ec          mov     dword ptr [ebp-14h],eax                       ; store waiters list ptr (old head) in local wait block

ntos!ExfAcquirePushLockShared+0x3e:
004c478e 8d45dd          lea     eax,[ebp-23h]                                 ; ptr to wait block ([ebp-24h]) + waiters bit set (1) 
004c4791 8bd0            mov     edx,eax
004c4793 8bc3            mov     eax,ebx
004c4795 f00fb116        lock cmpxchg dword ptr [esi],edx                      ; push local wait block onto waiters list, waiters bit set
004c4799 3bc3            cmp     eax,ebx
004c479b 7524            jne     ntos!ExfAcquirePushLockShared+0x71 (004c47c1)

ntos!ExfAcquirePushLockShared+0x4d:
004c479d 51              push    ecx                                           ; local wait block pushed to waiters list, wait for wake
004c479e 51              push    ecx
004c479f 51              push    ecx
004c47a0 6a1c            push    1Ch
004c47a2 8d45dc          lea     eax,[ebp-24h]
004c47a5 50              push    eax
004c47a6 e805b1f4ff      call    ntos!KeWaitForMutexObject (0040f8b0)          ; wait for KMUTEX on local wait block to be signalled
                                                                               ; NB: wait from wake => lock owned, fallthrough to exit
ntos!ExfAcquirePushLockShared+0x5b:
004c47ab 5e              pop     esi                                           ; exit, owning lock shared
004c47ac 5b              pop     ebx
004c47ad c9              leave
004c47ae c3              ret

ntos!ExfAcquirePushLockShared+0x5f:
004c47af 894dec          mov     dword ptr [ebp-14h],ecx                       ; clear waiters list ptr in local wait block (first waiter, old head = NULL)
004c47b2 ebda            jmp     ntos!ExfAcquirePushLockShared+0x3e (004c478e) ; jump to push local wait block to lock

ntos!ExfAcquirePushLockShared+0x64:
004c47b4 8d5304          lea     edx,[ebx+4]                                   ; increment share count (4)
004c47b7 8bc3            mov     eax,ebx
004c47b9 f00fb116        lock cmpxchg dword ptr [esi],edx                      ; perform (share count) -> (share count + 1) transition on lock
004c47bd 3bc3            cmp     eax,ebx
004c47bf 74ea            je      ntos!ExfAcquirePushLockShared+0x5b (004c47ab) ; jump to exit, owning lock shared, if transition succeeded

ntos!ExfAcquirePushLockShared+0x71:
004c47c1 8bd8            mov     ebx,eax                                        ; re-check lock state and retry if transition failed
004c47c3 eb99            jmp     ntos!ExfAcquirePushLockShared+0xe (004c475e)
0:000>
```
Repeating the exercise with Windows Server 2003 SP1 (`5.2.3790.1830 (srv03_sp1_rtm.050324-1447)`), we can see that the push lock code in the kernel matches the current structure of `EX_PUSH_LOCK` / `SRWLOCK`, and `ExfAcquirePushLockShared` has the anti-convoy shared -> exclusive upgrade path:
```
C:\Temp>curl -Lo ntos.ex_ https://archive.org/download/WinServer2003SP1AIO/2k3sp18in1.iso/STDC/NTKRNLMP.EX_
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 1122k    0 1122k    0     0   149k      0 --:--:--  0:00:07 --:--:--  194k

C:\Temp>expand ntos.ex_ ntos.exe
Microsoft (R) File Expansion Utility
Copyright (c) Microsoft Corporation. All rights reserved.

Adding C:\Temp\ntos.exe to Extraction Queue

Expanding Files ....

Expanding Files Complete ...

C:\Temp>cdb -z ntos.exe

Microsoft (R) Windows Debugger Version 10.0.22621.755 X86
Copyright (c) Microsoft Corporation. All rights reserved.


Loading Dump File [C:\Temp\ntos.exe]

************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
Symbol search path is: srv*
Executable search path is:
ModLoad: 00400000 00675000   C:\Temp\ntos.exe
eax=00000000 ebx=00000000 ecx=00000000 edx=00000000 esi=00000000 edi=00000000
eip=006211fc esp=00000000 ebp=00000000 iopl=0         nv up di pl nz na po nc
cs=0000  ss=0000  ds=0000  es=0000  fs=0000  gs=0000             efl=00000000
ntos!KiSystemStartup:
006211fc 55              push    ebp

0:000> lm vm nt*
start    end        module name
00400000 00675000   ntos       (pdb symbols)          C:\sym\ntkrnlmp.pdb\D11928D0A382448AA96E545C81E4DD172\ntkrnlmp.pdb
    Loaded symbol image file: ntos.exe
    Mapped memory image file: C:\Temp\ntos.exe
    Image path: C:\Temp\ntos.exe
    Image name: ntos.exe
    Timestamp:        Fri Mar 25 00:42:08 2005 (42435E60)
    CheckSum:         0025EC4B
    ImageSize:        00275000
    File version:     5.2.3790.1830
    Product version:  5.2.3790.1830
    File flags:       0 (Mask 3F)
    File OS:          40004 NT Win32
    File type:        1.0 App
    File date:        00000000.00000000
    Translations:     0409.04b0
    Information from resource tables:
        CompanyName:      Microsoft Corporation
        ProductName:      Microsoft® Windows® Operating System
        InternalName:     ntkrnlmp.exe
        OriginalFilename: ntkrnlmp.exe
        ProductVersion:   5.2.3790.1830
        FileVersion:      5.2.3790.1830 (srv03_sp1_rtm.050324-1447)
        FileDescription:  NT Kernel & System
        LegalCopyright:   © Microsoft Corporation. All rights reserved.

0:000> dt -v ntos!_EX_PUSH_LOCK
struct _EX_PUSH_LOCK, 7 elements, 0x4 bytes
   +0x000 Locked           : Bitfield Pos 0, 1 Bit
   +0x000 Waiting          : Bitfield Pos 1, 1 Bit
   +0x000 Waking           : Bitfield Pos 2, 1 Bit
   +0x000 MultipleShared   : Bitfield Pos 3, 1 Bit
   +0x000 Shared           : Bitfield Pos 4, 28 Bits
   +0x000 Value            : Uint4B
   +0x000 Ptr              : Ptr32 to Void

0:000> uf ntos!ExfAcquirePushLockShared
ntos!ExfAcquirePushLockShared+0xf7:                                             ; NB: anti-convoy shared -> exclusive upgrade path
00413ed6 8d4a01          lea     ecx,[edx+1]                                    ;  set lock bit (1)
00413ed9 e9c75d0100      jmp     ntos!ExfAcquirePushLockShared+0x102 (00429ca5) ;  jump to perform (!locked) -> (locked) transition
...
ntos!ExfAcquirePushLockShared:
00429c4e 8bff            mov     edi,edi                                        ; * acquire shared routine entry *
00429c50 53              push    ebx
00429c51 8bdc            mov     ebx,esp
00429c53 51              push    ecx
00429c54 51              push    ecx
00429c55 83e4f0          and     esp,0FFFFFFF0h
00429c58 83c404          add     esp,4
00429c5b 55              push    ebp
00429c5c 8b6b04          mov     ebp,dword ptr [ebx+4]
00429c5f 896c2404        mov     dword ptr [esp+4],ebp
00429c63 8bec            mov     ebp,esp
00429c65 83ec48          sub     esp,48h
00429c68 8365f000        and     dword ptr [ebp-10h],0
00429c6c 56              push    esi
00429c6d 57              push    edi
00429c6e 8bf1            mov     esi,ecx
00429c70 8b16            mov     edx,dword ptr [esi]                           ; load lock value -> edx
00429c72 33c0            xor     eax,eax
00429c74 8d7df4          lea     edi,[ebp-0Ch]
00429c77 ab              stos    dword ptr es:[edi]

ntos!ExfAcquirePushLockShared+0x2a:
00429c78 f6c201          test    dl,1                                          ; check for lock (1)
00429c7b 7419            je      ntos!ExfAcquirePushLockShared+0xf2 (00429c96) ; if not lock set, go to path to acquire lock

...                                                                            ; lock already held => increment share count if uncontended
                                                                               ; shared ownership, or build local wait block and enqueue
                                                                               ; to lock if held exclusive / lock under contention

ntos!ExfAcquirePushLockShared+0xf2:
00429c96 f6c202          test    dl,2                                          ; acquire path: check for waiters (2) (and no lock)
00429c99 0f8537a2feff    jne     ntos!ExfAcquirePushLockShared+0xf7 (00413ed6) ; if waiters and no lock, go to +0xf7 (anti-convoy shared -> exclusive upgrade)

ntos!ExfAcquirePushLockShared+0xfc:                                            ; no lock and no waiters:
00429c9f 8d4a10          lea     ecx,[edx+10h]                                 ;  increment share count (16)
00429ca2 83c901          or      ecx,1                                         ;  set lock bit (1)
                                                                               ; fall through to perform (!locked) -> (shared locked / share count = 1) transition
ntos!ExfAcquirePushLockShared+0x102:
00429ca5 8bc2            mov     eax,edx
00429ca7 f00fb10e        lock cmpxchg dword ptr [esi],ecx                      ; perform (!locked) -> ([shared] locked) transition
00429cab 3bc2            cmp     eax,edx
00429cad 0f859686ffff    jne     ntos!ExfAcquirePushLockShared+0x10c (00422349)

ntos!ExfAcquirePushLockShared+0x11e:
00429cb3 5f              pop     edi
00429cb4 5e              pop     esi
00429cb5 8be5            mov     esp,ebp
00429cb7 5d              pop     ebp
00429cb8 8be3            mov     esp,ebx
00429cba 5b              pop     ebx
00429cbb c3              ret
```
Though the `EX_PUSH_LOCK` functionality was exported from the kernel in Windows Server 2003 RTM, as far as I can tell it was not published as an API (either directly or indirectly) until Windows Server 2003 SP1 / Windows XP SP2, and was only consumed internally.

In Windows Server 2003 SP1 / Windows XP SP2, the kernel mode file system [filter manager](https://learn.microsoft.com/en-us/windows-hardware/drivers/ifs/filter-manager-concepts) (`fltmgr.sys`) was added, which duplicated the older (non convoy-avoiding) push lock code from Server 2003 RTM into `fltmgr.sys`, exposing thin wrappers around the `EX_PUSH_LOCK` functionality (e.g. entering a critical region using [`KeEnterCriticalRegion`](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntddk/nf-ntddk-keentercriticalregion) before acquiring a push lock by calling its copy of the older `ExfAcquirePushLock*` routines, etc.). This was the first time that `EX_PUSH_LOCK` functionality was exposed to 3rd-party drivers (albeit indirectly):
```
C:\Temp>curl -Lo fltmgr.sy_ https://archive.org/download/WinServer2003SP1AIO/2k3sp18in1.iso/ROOT/STDC/I386/FLTMGR.S
Y_
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 63205    0 63205    0     0  32291      0 --:--:--  0:00:01 --:--:-- 2168k

C:\Temp>expand fltmgr.sy_ flt.sys
Microsoft (R) File Expansion Utility
Copyright (c) Microsoft Corporation. All rights reserved.

Adding C:\Temp\flt.sys to Extraction Queue

Expanding Files ....

Expanding Files Complete ...

C:\Temp>cdb -z flt.sys

Microsoft (R) Windows Debugger Version 10.0.22621.755 X86
Copyright (c) Microsoft Corporation. All rights reserved.


Loading Dump File [C:\Temp\flt.sys]

************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
Symbol search path is: srv*
Executable search path is:
ModLoad: 00010000 00035000   C:\Temp\flt.sys
eax=00000000 ebx=00000000 ecx=00000000 edx=00000000 esi=00000000 edi=00000000
eip=00030ada esp=00000000 ebp=00000000 iopl=0         nv up di pl nz na po nc
cs=0000  ss=0000  ds=0000  es=0000  fs=0000  gs=0000             efl=00000000
flt!GsDriverEntry:
00030ada 8bff            mov     edi,edi

0:000> lm vm flt*
start    end        module name
00010000 00035000   flt        (pdb symbols)          C:\sym\fltMgr.pdb\CAD9573F320949DFB5054EE848988B251\fltMgr.pdb
    Loaded symbol image file: flt.sys
    Mapped memory image file: C:\Temp\flt.sys
    Image path: C:\Temp\flt.sys
    Image name: flt.sys
    Timestamp:        Fri Mar 25 00:30:25 2005 (42435BA1)
    CheckSum:         000221C6
    ImageSize:        00025000
    File version:     5.2.3790.1830
    Product version:  5.2.3790.1830
    File flags:       0 (Mask 3F)
    File OS:          40004 NT Win32
    File type:        3.7 Driver
    File date:        00000000.00000000
    Translations:     0409.04b0
    Information from resource tables:
        CompanyName:      Microsoft Corporation
        ProductName:      Microsoft® Windows® Operating System
        InternalName:     fltMgr.sys
        OriginalFilename: fltMgr.sys
        ProductVersion:   5.2.3790.1830
        FileVersion:      5.2.3790.1830 (srv03_sp1_rtm.050324-1447)
        FileDescription:  Microsoft Filesystem Filter Manager
        LegalCopyright:   © Microsoft Corporation. All rights reserved.

0:000> uf flt!FltAcquirePushLockShared
flt!FltAcquirePushLockShared:
00015ac2 8bff            mov     edi,edi
00015ac4 55              push    ebp
00015ac5 8bec            mov     ebp,esp
00015ac7 56              push    esi
00015ac8 ff152c900100    call    dword ptr [flt!_imp__KeEnterCriticalRegion (0001902c)] ; enter critical region
00015ace 8b4d08          mov     ecx,dword ptr [ebp+8]
00015ad1 8b11            mov     edx,dword ptr [ecx]
00015ad3 83e2fc          and     edx,0FFFFFFFCh                                         ; build lock value with (share count + 1)
00015ad6 8d7204          lea     esi,[edx+4]
00015ad9 8bc2            mov     eax,edx
00015adb f00fb131        lock cmpxchg dword ptr [ecx],esi                               ; try acquire with (share count) -> (share count + 1)
00015adf 3bc2            cmp     eax,edx
00015ae1 5e              pop     esi
00015ae2 7405            je      flt!FltAcquirePushLockShared+0x27 (00015ae9)

flt!FltAcquirePushLockShared+0x22:
00015ae4 e8df2b0000      call    flt!FltpExfAcquirePushLockShared (000186c8)            ; if that fails, fall back to full copy of old ExfAcquirePushLockShared code

flt!FltAcquirePushLockShared+0x27:
00015ae9 5d              pop     ebp
00015aea c20400          ret     4

0:000> uf flt!FltpExfAcquirePushLockShared
flt!FltpExfAcquirePushLockShared:                                                       ; this works without convoy avoidance, as per Server 2003 RTM code above
000186c8 8bff            mov     edi,edi
000186ca 55              push    ebp                                                    ; [there's a small tweak vs. the 2003 RTM code to use KEVENT instead of
000186cb 8bec            mov     ebp,esp                                                ;  KMUTEX as the wait object in the case you have to block on the lock,
000186cd 83ec24          sub     esp,24h                                                ;  but otherwise the logic is unchanged]
000186d0 53              push    ebx
000186d1 56              push    esi
000186d2 8bf1            mov     esi,ecx
000186d4 8b1e            mov     ebx,dword ptr [esi]
000186d6 57              push    edi

flt!FltpExfAcquirePushLockShared+0xf:
000186d7 f6c303          test    bl,3
000186da 750f            jne     flt!FltpExfAcquirePushLockShared+0x23 (000186eb)

flt!FltpExfAcquirePushLockShared+0x14:
000186dc 8d4b04          lea     ecx,[ebx+4]
000186df 8bc3            mov     eax,ebx
000186e1 f00fb10e        lock cmpxchg dword ptr [esi],ecx
000186e5 3bc3            cmp     eax,ebx
000186e7 7450            je      flt!FltpExfAcquirePushLockShared+0x71 (00018739)

flt!FltpExfAcquirePushLockShared+0x21:
000186e9 eb3b            jmp     flt!FltpExfAcquirePushLockShared+0x5e (00018726)

flt!FltpExfAcquirePushLockShared+0x23:
000186eb 33ff            xor     edi,edi
000186ed 57              push    edi
000186ee 6a01            push    1
000186f0 8d45dc          lea     eax,[ebp-24h]
000186f3 50              push    eax
000186f4 ff1548900100    call    dword ptr [flt!_imp__KeInitializeEvent (00019048)]
000186fa f6c301          test    bl,1
000186fd c645fc00        mov     byte ptr [ebp-4],0
00018701 897df8          mov     dword ptr [ebp-8],edi
00018704 897df0          mov     dword ptr [ebp-10h],edi
00018707 897df4          mov     dword ptr [ebp-0Ch],edi
0001870a 7408            je      flt!FltpExfAcquirePushLockShared+0x4c (00018714)

flt!FltpExfAcquirePushLockShared+0x44:
0001870c 8d43ff          lea     eax,[ebx-1]
0001870f 8945ec          mov     dword ptr [ebp-14h],eax
00018712 eb03            jmp     flt!FltpExfAcquirePushLockShared+0x4f (00018717)

flt!FltpExfAcquirePushLockShared+0x4c:
00018714 897dec          mov     dword ptr [ebp-14h],edi

flt!FltpExfAcquirePushLockShared+0x4f:
00018717 8d45dd          lea     eax,[ebp-23h]
0001871a 8bc8            mov     ecx,eax
0001871c 8bc3            mov     eax,ebx
0001871e f00fb10e        lock cmpxchg dword ptr [esi],ecx
00018722 3bc3            cmp     eax,ebx
00018724 7404            je      flt!FltpExfAcquirePushLockShared+0x62 (0001872a)

flt!FltpExfAcquirePushLockShared+0x5e:
00018726 8bd8            mov     ebx,eax
00018728 ebad            jmp     flt!FltpExfAcquirePushLockShared+0xf (000186d7)

flt!FltpExfAcquirePushLockShared+0x62:
0001872a 57              push    edi
0001872b 57              push    edi
0001872c 57              push    edi
0001872d 6a1c            push    1Ch
0001872f 8d45dc          lea     eax,[ebp-24h]
00018732 50              push    eax
00018733 ff1570900100    call    dword ptr [flt!_imp__KeWaitForSingleObject (00019070)]

flt!FltpExfAcquirePushLockShared+0x71:
00018739 5f              pop     edi
0001873a 5e              pop     esi
0001873b 5b              pop     ebx
0001873c c9              leave
0001873d c3              ret
0:000>
```
Filter manager (`fltmgr.sys`) retained its own version of the push lock code until Windows Vista RTM, at which point it was updated to using the push lock code exported from the kernel. This would have provided both performance and maintenance benefits, as the filter manager then obtained the anti-convoy lock features from the push lock implementation provided the kernel and a duplicate (and diverged) push lock implementation no longer needed to be maintained in the filter manager itself.

From Vista RTM onwards (to the present day), the filter manager has continued to use the push lock code provided by the kernel:
```
0:000> lm vm flt*
start             end                 module name
00000000`00010000 00000000`00053000   fltMgr     (pdb symbols)          c:\sym\fltMgr.pdb\A35A4E9939754866A8FB3EC3B80C83772\fltMgr.pdb
    Loaded symbol image file: fltMgr.sys
    Mapped memory image file: C:\Temp\fltMgr.sys
    Image path: C:\Temp\fltMgr.sys
    Image name: fltMgr.sys
    Timestamp:        Thu Nov  2 09:01:19 2006 (4549B3DF)
    CheckSum:         00048A13
    ImageSize:        00043000
    File version:     6.0.6000.16386
    Product version:  6.0.6000.16386
    File flags:       0 (Mask 3F)
    File OS:          40004 NT Win32
    File type:        3.7 Driver
    File date:        00000000.00000000
    Translations:     0409.04b0
    Information from resource tables:
        CompanyName:      Microsoft Corporation
        ProductName:      Microsoft® Windows® Operating System
        InternalName:     fltMgr.sys
        OriginalFilename: fltMgr.sys
        ProductVersion:   6.0.6000.16386
        FileVersion:      6.0.6000.16386 (vista_rtm.061101-2205)
        FileDescription:  Microsoft Filesystem Filter Manager
        LegalCopyright:   © Microsoft Corporation. All rights reserved.

0:000> uf fltMgr!FltAcquirePushLockShared
fltMgr!FltAcquirePushLockShared:
00000000`00012820 fff3            push    rbx
00000000`00012822 4883ec20        sub     rsp,20h
00000000`00012826 488bd9          mov     rbx,rcx
00000000`00012829 ff1589e80000    call    qword ptr [fltMgr!_imp_KeEnterCriticalRegion (00000000`000210b8)] ; enter critical region
00000000`0001282f 33c0            xor     eax,eax
00000000`00012831 41bb11000000    mov     r11d,11h                                    
00000000`00012837 f04c0fb11b      lock cmpxchg qword ptr [rbx],r11                                          ; try (unlocked) -> (lock / share count = 1) transition to acquire shared lock 
00000000`0001283c 0f85b82c0000    jne     fltMgr! ?? ::FNODOBFM::`string'+0x890 (00000000`000154fa)

fltMgr!FltAcquirePushLockShared+0x22:
00000000`00012842 4883c420        add     rsp,20h
00000000`00012846 5b              pop     rbx
00000000`00012847 c3              ret

fltMgr! ?? ::FNODOBFM::`string'+0x890:
00000000`000154fa 488bcb          mov     rcx,rbx
00000000`000154fd ff154dbc0000    call    qword ptr [fltMgr!_imp_ExfAcquirePushLockShared (00000000`00021150)] ; try acquire failed, call the real acquire shared routine in the kernel
00000000`00015503 90              nop
00000000`00015504 e939d3ffff      jmp     fltMgr!FltAcquirePushLockShared+0x22 (00000000`00012842)
```
It was also in Windows Vista RTM that the user-mode `SRWLOCK` counterpart for `EX_PUSH_LOCK` was released. However, despite the improvements to the filter manager in Vista RTM, push locks were still not an officially-supported mechanism for 3rd-party drivers until Windows 8(!). At that point, official `Ex` APIs for push locks were added and drivers could use them directly (rather than via the filter manager) &mdash; cf. these comments from Alex Ionescu in an OSR article:

> A final remark here is that officially, Pushlock support for 3rd party driver developers is only available starting in Windows 8 and later. Previously, accessing this lock required linking with the File System Mini-Filter Manager and using the FltXxx routines that wrap the Pushlock API, which might look odd for say, a network card driver.
> 
> https://www.osr.com/nt-insider/2015-issue3/the-state-of-synchronization/

Windows 8 also contained work from me to further improve the convoy-avoidance of push locks by extending the two-phase unlock work @NeillClift had done in Server 2003 SP1.

Taking all this together, the `SRWLOCK` code has always, since its release in Windows Vista, contained convoy avoidance and the shared -> exclusive upgrade path. The kernel-mode equivalents, in the form of `EX_PUSH_LOCK` in the kernel, has had the same anti-convoy logic since Windows Server 2003 SP1, though the filter manager's duplicate copy of the push lock functionality retained the non-convoy-avoiding prior push lock logic from Windows Server 2003 SP1 until Windows Vista.

Since Windows Vista RTM (and Windows Server 2008 RTM), then, *all* user-mode and *all* kernel-mode callers of this code have had the benefit anti-convoy logic, and the performance improvements that accrue from that. Given the discussion above, it's clear that the Windows team will have to provide an option for performing a shared acquire on `SRWLOCK` that will not perform a shared -> exclusive upgrade, so that the MS STL can be implemented with the semantics that the C++ spec demands.

However, the C++ STL is far from the only user of `SRWLOCK` and (clearly!) is not a user of `EX_PUSH_LOCK` in kernel mode. The Windows team will not have control of many of the callers of these APIs, and so unilaterally removing the shared -> exclusive upgrade path could damage performance for such callers, with no easy way for that to be remedied. The Windows team would then have to provide an option to opt-in to using the shared -> exclusive upgrade path, and ask the callers whose performance had been harmed to update and re-release their software using the new API/option. While it's true that MS have control of any software of their own which calls these APIs (whether within the Windows team, or in other products), even for callers within MS software outside the Windows team this poses a large burden on creating such updates. Furthermore, the performance of the software using these is adversely impacted until they can do this. This problem is magnified when it comes to other software vendors, who then have to be made aware of this problem and encouraged to migrate to a new API/option, just to get the behaviour which they had before.

I don't have sight of the performance numbers, but if the performance hit is big enough then in my mind making a unilateral change that adversely affects performance of other software, without a source change and re-release, would be a big demand to make of customers outside the Windows team (whether internal to MS, or external). Given https://github.com/MicrosoftDocs/sdk-api/pull/1785 has documented that a potential shared -> exclusive upgrade is the expected behaviour of `SRWLOCK`, I think that the better option would simply be for the Windows team to provide a new API/option for the C++ STL to use to avoid the shared -> exclusive upgrade path, so that `shared_mutex` gets the semantics it expects.

It is my suspicion that any change to the `SRWLOCK` won't meet the servicing bar for most older versions of Windows, and so any change (whether unilateral or not) will likely occur only in the latest Win 11 version. If that is the case, then either the C++ STL will only be able to have a fix on the latest Windows version, or it would have to implement `shared_mutex` on its own without using Win32's `SRWLOCK`. I have provided a C++17-compliant workaround using condition variables, both here and on the [StackOverflow post](https://stackoverflow.com/questions/78090862/stdshared-mutexunlock-shared-blocks-even-though-there-are-no-active-exclus/78137800#78137800), and so if the C++ STL continues to use `SRWLOCK` and a fix is only available with the latest Windows, then this issue would effectively be a 'wontfix' for such older platforms, but with the small positive of a C++17 standard-compliant workaround being available and tested.

Given that relatively likely outcome, and the fact that the potential shared -> exclusive upgrade behaviour of `SRWLOCK` is now documented, I think this adds further weight to the argument not to unilaterally change `SRWLOCK` (or `EX_PUSH_LOCK`), as a unilateral change in that case (without backports to all Windows versions that the C++ STL supports) would not provide a complete solution to this issue and would detriment many users outside the C++ STL.

**Comment by adr26** [comment]
>  I think the code severely impacted by lock convoys should instead fix its lock usage, like reducing parallelism, reducing locks scope, going lock-free where applicable, etc, rather that relying on that the lock itself would automatically fix their problem.

The crux of the problem here is that you can't necessarily tell ahead of time what code is going to get impacted by lock convoys &mdash; if a lock is fair, then the lock will end up getting held over context switches as the lock is passed in a fair manner from one waiter to the next, if the waiter isn't switched in. Convoys are then caused as an artifact of this, even in cases where the caller is well-behaved and has short lock hold times, because the callers have no control over OS features such as the context switch time!

Since convoys can (and do) occur for well-behaved callers, in a manner that is out of their control, then it's appropriate for a lock to help avoid convoys &mdash; as long as the fact that the lock will no longer exhibit fair behaviour (or other surprising quirks such as the shared -> exclusive upgrade path!) are documented &mdash; in order to improve system performance.

The C++ spec is layering a further requirement on top of `SRWLOCK` that precludes one of the anti-convoy techniques it's currently using (the shared -> exclusive upgrade path), and which it has used throughout the entirety of its existence. Allowing the C++ STL to specify that it doesn't want this behaviour, without precluding the other anti-convoy behaviour of the `SRWLOCK` (unfair exclusive ownership stealing by exclusive acquirers), would allow the C++ STL to have the behaviour that the C++ spec demands at the minimal performance cost to the C++ STL while also having no performance impact on other callers with respect to what is a valid optimisation.

It is therefore what I would equally strongly (and conversely) recommend as the fix!

**Comment by tialaramex** [comment]
Is there a way to track the progress and (hopefully) eventual resolution of the SRWLOCK bug independent of this STL issue? For Rust's stdlib it would be nice to understand whether Microsoft intends to fix this, and how because of course the Rust stdlib made the same assumption as the STL. The OS bug link is Microsoft internal, but maybe there's a public view into that ?

**Comment by adr26** [comment]
> Is there a way to track the progress and (hopefully) eventual resolution of the SRWLOCK bug independent of this STL issue? For Rust's stdlib it would be nice to understand whether Microsoft intends to fix this, and how because of course the Rust stdlib made the same assumption as the STL. The OS bug link is Microsoft internal, but maybe there's a public view into that ?

I can't comment on the internal OS bug (as I don't have access), but for future reference I see that Rust documents the type [std](https://doc.rust-lang.org/std/index.html)::[sync](https://doc.rust-lang.org/std/sync/index.html)::[RwLock](https://doc.rust-lang.org/std/sync/struct.RwLock.html#) as follows (emphasis mine):

> In comparison, a [`Mutex`](https://doc.rust-lang.org/std/sync/struct.Mutex.html) does not distinguish between readers or writers that acquire the lock, therefore blocking any threads waiting for the lock to become available. **An `RwLock` will allow any number of readers to acquire the lock as long as a writer is not holding the lock.**

https://doc.rust-lang.org/std/sync/struct.RwLock.html

In the Windows implementation of Rust, `SRWLOCK` is used as the underlying lock type and the `RwLock` `read()` API is implemented using a call to `AcquireSRWLockShared()` on the `SRWLOCK`:

https://github.com/rust-lang/rust/blob/01f7f3a1ff983ddfe485bbd9cedd59c378fdaceb/library/std/src/sys/sync/rwlock/windows7.rs#L18

This results in a conflict between the (now documented) behaviour of `AcquireSRWLockShared()` and the Rust requirement that an unbounded number of concurrent shared acquisitions may be performed when a thread holds the `RwLock` as a reader

**Comment by tialaramex** [comment]
I'm aware of how Mara's implementation of RwLock on Windows works which is why I asked my original question. The only new information to me in your response is the claim that this bug is "now documented" to which I'd ask: Where exactly did Microsoft document this extraordinary behaviour, and when?

**Comment by Alcaro** [comment]
It's supposed to be documented at https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-acquiresrwlockshared since about a month ago (https://github.com/MicrosoftDocs/sdk-api/pull/1785), but it seems that page is out of sync.

And that remark doesn't actually explain anything. Baffling behavior like this should be made abundantly clear, for example "An SRW lock may, in rare cases, limit its simultaneous shared mode acquisitions, potentially down to 1, for performance reasons."

**Comment by tialaramex** [comment]
Rust's stdlib will cease to rely on SRWLock once Rust 1.78 is published on Thursday 2 May 2024 so that's good news. Properly documenting that SRWLock has this very weird behaviour still seems like a good idea for those who aren't on Rust.

**Comment by StephanTLavavej** [comment]
Windows has been fixed with internal [OS-PR-10397182](https://microsoft.visualstudio.com/OS/_git/os.2020/pullrequest/10397182) "Disable implicit upgrade for PushLock and SRWLock", which has entirely removed the problematic behavior.

I don't know when this change will be released - at this point "a future release of Windows" is all that I can expect. And although I have no specific information, I believe that it would be highly unlikely for this change to be backported to any released versions of Windows ("downlevel").

**Comment by wnstngs** [comment]
@cpkleynhans @StephanTLavavej Does that mean the shared to exclusive upgrade path was removed? How are they handling the convoy problem now?

**Comment by Alcaro** [comment]
If this is (will be) fixed in Windows 12, isn't it still a problem as long as any older Windows version remains supported?

**Comment by StephanTLavavej** [comment]
@wnstngs I asked the author of the fix to comment.

@Alcaro If anyone (whether a programmer-user or an end-user) wants this fix, they should upgrade. (Again, I don't know anything about the ship vehicle - I would expect this to appear in a future update to Windows 11.) That's why software versions exist - upgrading to the latest version gives generally increasing correctness and performance. Staying on an older version gives behavioral stability, even at the expense of correctness.

Backporting fixes is inherently destabilizing. This is of course necessary for important things like security, critical regressions, and the like. But this SRWLOCK behavior can't be *that* bad (we have somehow survived with it for many years, including all of `std::shared_mutex`'s existence). It was bad enough to be worth fixing (yay) but again, I highly doubt that it will be considered bad enough to be worth backporting.

**Comment by vm2mv** [comment]
This issue is critical because some programs need powerful servers due to using std::shared_mutex.

If updating the OS isn’t possible, std::shared_mutex should be changed to stop using SRWLock and avoid deadlocks.
